### PR TITLE
Add support for client texture selection

### PR DESCRIPTION
**Pull request by huffman**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9002/app_12956_PR_bb24a061546531ad35c80808e178969fe6ca80e2.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9002/HighFidelity-Beta-PR12956-bb24a061546531ad35c80808e178969fe6ca80e2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9002/HighFidelity-Beta-Interface-PR12956-bb24a061546531ad35c80808e178969fe6ca80e2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9002/HighFidelity-Beta-PR12956-bb24a061546531ad35c80808e178969fe6ca80e2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9002/HighFidelity-Beta-Interface-PR12956-bb24a061546531ad35c80808e178969fe6ca80e2.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9006/app_12956_PR_b7e2723cad2fc4df9649ff5378d8e7506b806653.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9006/HighFidelity-Beta-PR12956-b7e2723cad2fc4df9649ff5378d8e7506b806653.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9006/HighFidelity-Beta-Interface-PR12956-b7e2723cad2fc4df9649ff5378d8e7506b806653.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9006/HighFidelity-Beta-PR12956-b7e2723cad2fc4df9649ff5378d8e7506b806653.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9009/app_12956_PR_ef46422610c45b4a219cb7c40d0481153eff161d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9009/HighFidelity-Beta-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9009/HighFidelity-Beta-Interface-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9009/HighFidelity-Beta-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9009/HighFidelity-Beta-Interface-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.exe

**Comment by huffman** [comment]
To test:

* https://highfidelity.testrail.net/index.php?/suites/view/58
* Add a model entity with the url https://s3.amazonaws.com/hifi-public/huffman/secretGardenParty.fbx
* Upload https://s3.amazonaws.com/hifi-public/huffman/secretGardenParty.fbx to your asset server.
  * Confirm it successfully bakes
  * Restart interface
  * Add the asset to the world from the Asset Browser
  * The model should fully load. Compare it to the model entity you previously added and make sure they look the same

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9010/app_12956_PR_ef46422610c45b4a219cb7c40d0481153eff161d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9010/HighFidelity-Beta-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9010/HighFidelity-Beta-Interface-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9010/HighFidelity-Beta-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9010/HighFidelity-Beta-Interface-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.exe

**Comment by huffman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9022/app_12956_PR_ef46422610c45b4a219cb7c40d0481153eff161d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9022/HighFidelity-Beta-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9022/HighFidelity-Beta-Interface-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9022/HighFidelity-Beta-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9022/HighFidelity-Beta-Interface-PR12956-ef46422610c45b4a219cb7c40d0481153eff161d.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9067/app_12956_PR_9b3c84e2dc5ac9c7de17c354e0e38c732db5407e.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9069/app_12956_PR_4980521e8cec1e3736717b33c7c89d044a35e2f4.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9069/HighFidelity-Beta-PR12956-4980521e8cec1e3736717b33c7c89d044a35e2f4.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9069/HighFidelity-Beta-Interface-PR12956-4980521e8cec1e3736717b33c7c89d044a35e2f4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9069/HighFidelity-Beta-PR12956-4980521e8cec1e3736717b33c7c89d044a35e2f4.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9069/HighFidelity-Beta-Interface-PR12956-4980521e8cec1e3736717b33c7c89d044a35e2f4.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Asset browser (When working) uploads and successfully displays as expected.

**Comment by conklin94122** [comment]
We haven't cut RC67 so this can't be merged in for RC68 yet.

**Comment by conklin94122** [comment]
@huffman Build conflicts

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9259/app_12956_PR_d96ee7a6db035a76a94a76cb948948b1065fac99.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9259/HighFidelity-Beta-PR12956-d96ee7a6db035a76a94a76cb948948b1065fac99.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9259/HighFidelity-Beta-Interface-PR12956-d96ee7a6db035a76a94a76cb948948b1065fac99.dmg

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Asset browser looks good and models uploaded to it bake and successfully are added to world.

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 1]
```diff
@@ -309,17 +310,25 @@ static bool isLocalUrl(const QUrl& url) {
...
     Resource(url),
     _type(type),
-    _sourceIsKTX(url.path().endsWith(".ktx")),
     _maxNumPixels(maxNumPixels)
 {
     _textureSource = std::make_shared<gpu::TextureSource>(url, (int)type);
     _lowestRequestedMipLevel = 0;
 
-    _shouldFailOnRedirect = !_sourceIsKTX;
+    qDebug() << "Creating networktexture: " << url;
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: Still want this?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 2]
```diff
@@ -309,17 +310,25 @@ static bool isLocalUrl(const QUrl& url) {
...
+    if (url.fileName().endsWith(TEXTURE_META_EXTENSION)) {
+        _currentlyLoadingResourceType = ResourceType::META;
+    } else if (url.fileName().endsWith(".ktx")) {
+        _currentlyLoadingResourceType = ResourceType::KTX;
+    } else {
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: Should the two above endsWith compares be case insensitive?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 3]
```diff
@@ -393,12 +402,13 @@ NetworkTexture::~NetworkTexture() {
 
 const uint16_t NetworkTexture::NULL_MIP_LEVEL = std::numeric_limits<uint16_t>::max();
 void NetworkTexture::makeRequest() {
-    if (!_sourceIsKTX) {
+    qDebug() << "In makeRequest for " << _activeUrl << (int)_currentlyLoadingResourceType;
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: still want this debug?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 4]
```diff
@@ -444,6 +454,7 @@ void NetworkTexture::makeRequest() {
 
         _ktxHeaderRequest->send();
 
+        qDebug() << "Starting mip range request";
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: still want this debug?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 5]
```diff
@@ -581,6 +594,7 @@ void NetworkTexture::startMipRangeRequest(uint16_t low, uint16_t high) {
 
     bool isHighMipRequest = low == NULL_MIP_LEVEL && high == NULL_MIP_LEVEL;
 
+    qDebug() << "Making ktx mip request to: " << _activeUrl;
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: still want this debug?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 6]
```diff
@@ -930,11 +944,79 @@ void NetworkTexture::handleFinishedInitialLoad() {
 }
 
 void NetworkTexture::downloadFinished(const QByteArray& data) {
-    loadContent(data);
+    qDebug() << "Loading content: " << _activeUrl;
+    if (_currentlyLoadingResourceType == ResourceType::META) {
+        qDebug() << "Loading meta content: " << _activeUrl;
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: Still want the two debug lines here?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 7]
```diff
@@ -930,11 +944,79 @@ void NetworkTexture::handleFinishedInitialLoad() {
...
+            if (backend->supportedTextureFormat(elFormat)) {
+                auto url = pair.second;
+                if (url.fileName().endsWith(TEXTURE_META_EXTENSION)) {
+                    qWarning() << "Found a texture meta URL inside of the texture meta file at" << _activeUrl;
+                    continue;
+                }
+
+                _currentlyLoadingResourceType = ResourceType::KTX;
+                _activeUrl = _activeUrl.resolved(url);
+                qDebug() << "Active url is now: " << _activeUrl;
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: Still need?

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 8]
```diff
@@ -930,11 +944,79 @@ void NetworkTexture::handleFinishedInitialLoad() {
...
+        auto self = _self.lock();
+        if (!self) {
+            return;
+        }
+        QMetaObject::invokeMethod(this, "attemptRequest", Qt::QueuedConnection);
+        return;
+    }
+
+    qWarning() << "Failed to find supported texture type in " << _activeUrl;
+    //TextureCache::requestCompleted(_self);
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: remove if unneeded

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 9]
```diff
@@ -930,11 +944,79 @@ void NetworkTexture::handleFinishedInitialLoad() {
...
+    Resource::handleFailedRequest(ResourceRequest::NotFound);
 }
 
-void NetworkTexture::loadContent(const QByteArray& content) {
-    if (_sourceIsKTX) {
+void NetworkTexture::loadTextureContent(const QByteArray& content) {
+    if (_currentlyLoadingResourceType != ResourceType::ORIGINAL) {
+        qWarning() << "Trying to load texture content when currentl resource type is not ORIGINAL";
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: when current

File: libraries/networking/src/ResourceCache.cpp [hunk 10]
```diff
@@ -618,7 +619,7 @@ void Resource::init(bool resetLoaded) {
         _loaded = false;
     }
     _attempts = 0;
-    _activeUrl = _url;
+    qDebug() << "Initting resource: " << _url;
```
Comment (birarda) [inline libraries/networking/src/ResourceCache.cpp]: still need this?

File: libraries/networking/src/ResourceCache.cpp [hunk 11]
```diff
@@ -671,6 +672,7 @@ void Resource::makeRequest() {
 
     PROFILE_ASYNC_BEGIN(resource, "Resource:" + getType(), QString::number(_requestID), { { "url", _url.toString() }, { "activeURL", _activeUrl.toString() } });
 
+    qDebug() << "Making request to " << _activeUrl;
```
Comment (birarda) [inline libraries/networking/src/ResourceCache.cpp]: stil need this?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9372/HighFidelity-Beta-PR12956-12d4cf12cf6790a5d9549df29bc3718648a32994.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9372/HighFidelity-Beta-Interface-PR12956-12d4cf12cf6790a5d9549df29bc3718648a32994.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9372/HighFidelity-Beta-PR12956-12d4cf12cf6790a5d9549df29bc3718648a32994.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9372/HighFidelity-Beta-Interface-PR12956-12d4cf12cf6790a5d9549df29bc3718648a32994.exe

File: libraries/baking/src/ModelBaker.cpp [hunk 12]
```diff
@@ -519,3 +519,95 @@ void ModelBaker::setWasAborted(bool wasAborted) {
...
+                    videoNode.children.append({ "RelativeFilename", { relativeFilename }, { } });
+                    videoNode.children.append({ "Content", { textureFile.readAll() }, { } });
+
+                    rootChild.children.append(videoNode);
+
+                    textureFile.close();
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (huffman) [reply libraries/baking/src/ModelBaker.cpp]: The FBXBaker version was removed. Good catch!

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9439/app_12956_PR_529869e80c79bd281d60e526d0a3fce219132c55.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9439/HighFidelity-Beta-PR12956-529869e80c79bd281d60e526d0a3fce219132c55.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9439/HighFidelity-Beta-Interface-PR12956-529869e80c79bd281d60e526d0a3fce219132c55.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9492/app_12956_PR_529869e80c79bd281d60e526d0a3fce219132c55.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9492/HighFidelity-Beta-PR12956-529869e80c79bd281d60e526d0a3fce219132c55.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9492/HighFidelity-Beta-Interface-PR12956-529869e80c79bd281d60e526d0a3fce219132c55.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9492/HighFidelity-Beta-PR12956-529869e80c79bd281d60e526d0a3fce219132c55.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9492/HighFidelity-Beta-Interface-PR12956-529869e80c79bd281d60e526d0a3fce219132c55.exe

File: libraries/model-networking/src/model-networking/TextureCache.cpp [hunk 13]
```diff
@@ -309,17 +310,25 @@ static bool isLocalUrl(const QUrl& url) {
...
     _maxNumPixels(maxNumPixels)
 {
     _textureSource = std::make_shared<gpu::TextureSource>(url, (int)type);
     _lowestRequestedMipLevel = 0;
 
-    _shouldFailOnRedirect = !_sourceIsKTX;
+    auto fileNameLowercase = url.fileName().toLower();
+    if (fileNameLowercase.endsWith(TEXTURE_META_EXTENSION)) {
+        _currentlyLoadingResourceType = ResourceType::META;
+    } else if (fileNameLowercase.endsWith(".ktx")) {
```
Comment (birarda) [inline libraries/model-networking/src/model-networking/TextureCache.cpp]: To be safe these should be case insensitive, right?

File: libraries/baking/src/ModelBaker.cpp [hunk 14]
```diff
@@ -248,7 +248,7 @@ QString ModelBaker::compressTexture(QString modelTextureFileName, image::Texture
 
     QFileInfo modelTextureFileInfo{ modelTextureFileName.replace("\\", "/") };
     
-    if (modelTextureFileInfo.suffix() == BAKED_TEXTURE_EXT.mid(1)) {
+    if (modelTextureFileInfo.suffix() == BAKED_TEXTURE_KTX_EXT.mid(1)) {
```
Comment (birarda) [inline libraries/baking/src/ModelBaker.cpp]: if we're checking case insensitively when requesting I guess we should be doing that here too

File: libraries/model-networking/src/model-networking/TextureCache.cpp (hunk 13 above)
Comment (huffman) [reply libraries/model-networking/src/model-networking/TextureCache.cpp]: Hmm, aren't they?

Comment (birarda) [reply libraries/model-networking/src/model-networking/TextureCache.cpp]: Yes, I am stupid

File: libraries/ktx/src/TextureMeta.cpp [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+#include "TextureMeta.h"
+
+#include <QJsonDocument>
+#include <QJsonObject>
+
+const QString TEXTURE_META_EXTENSION = ".texmeta.json";
+
+bool TextureMeta::deserialize(const QByteArray& data, TextureMeta* meta) {
+    QJsonParseError error;
+    auto doc = QJsonDocument::fromJson(data, &error);
```
[the remark this answers is not in the document]
Comment (huffman) [reply libraries/ktx/src/TextureMeta.cpp]: Good catch, i added it to the if

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9503/app_12956_PR_e1d218bb57d11ed2b83003ab005dbc6fcb30f6b9.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9503/HighFidelity-Beta-PR12956-e1d218bb57d11ed2b83003ab005dbc6fcb30f6b9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9503/HighFidelity-Beta-Interface-PR12956-e1d218bb57d11ed2b83003ab005dbc6fcb30f6b9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9503/HighFidelity-Beta-PR12956-e1d218bb57d11ed2b83003ab005dbc6fcb30f6b9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9503/HighFidelity-Beta-Interface-PR12956-e1d218bb57d11ed2b83003ab005dbc6fcb30f6b9.exe

**Comment by huffman** [comment]
@qlimnavarro Would you be able to retest this?

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Asset browser works as expected.

**Pull request merged**